docs/design-documents/features/storage/SecureStore/SecureStore_design.md

```diff
@@ -0,0 +1,448 @@
+# SecureStore in Mbed OS
+
+- [SecureStore in Mbed OS](#securestore-in-mbed-os)
+    + [Revision history](#revision-history)
+- [Introduction](#introduction)
+    + [Overview and background](#overview-and-background)
+    + [Requirements and assumptions](#requirements-and-assumptions)
+- [System architecture and high-level design](#system-architecture-and-high-level-design)
+  * [Design basics](#design-basics)
+    + [Data layout](#data-layout)
+    + [Basic implementation concepts](#basic-implementation-concepts)
+- [Detailed design](#detailed-design)
+    + [Class header](#class-header)
+    + [Important data structures](#important-data-structures)
+    + [Initialization and reset](#initialization-and-reset)
+    + [Core APIs](#core-apis)
+    + [Incremental set APIs](#incremental-set-apis)
+    + [Key iterator APIs](#key-iterator-apis)
+- [Usage scenarios and examples](#usage-scenarios-and-examples)
+    + [Standard usage of the class](#standard-usage-of-the-class)
+- [Other information](#other-information)
+    + [Open issues](#open-issues)
+
+
+### Revision history
+
+| Revision 	| Date           	| Authors                                                	| Mbed OS version 	| Comments         	|
+|----------	|----------------	|--------------------------------------------------------	|-----------------	|------------------	|
+| 1.0      	| 02 October 2018	| David Saada ([@davidsaada](https://github.com/davidsaada/)) 	| 5.11+           	| Initial revision 	|
+
+# Introduction
+
+### Overview and background
+
+SecureStore is a [KVStore](../KVStore/KVStore_design.md) based storage solution, providing security features on the stored data, such as encryption, authentication, rollback protection and write once, over an underlying KVStore class. It references an additional KVStore class for storing the rollback protection keys. 
+
+### Requirements and assumptions
+
+SecureStore assumes that the underlying KVStore instances are instantiated and initialized. 
+
+# System architecture and high-level design
+
+## Design basics
+
+SecureStore is a storage class, derived from KVStore. It adds security features to the underlying key value store.
+ 
+As such, it offers all KVStore APIs, with additional security options (which can be selected using the creation flags at set). These include:  
+- Encryption: Data is encrypted using the AES-CTR encryption method, where the IV consists of a randomly generated 8-byte nonce and an 8-byte counter. Key is derived from [Device Key](../../../../../../mbed-os/features/device_key/README.md), where salt is the key trimmed to 32 bytes, with "ENC" as prefix.  
```

Comment:
What algorithm is used for the key derivation?

Reply:
We simply use the key derivation API `get_derived_key` from the `DeviceKey` feature, which implements the NIST SP 800-108 KDF in counter mode spec.

Reply:
cool